### PR TITLE
HTBHF-661 Add HMRC endpoint to the smart stub.

### PR DESCRIPTION
**Pull request by mikeSmith08**
- Added new controller for HMRC plus HMRC request object
 - Modified benefit service to deal with both DWP and HMRC
 - Added extra tests for HMRC in service and controller
 - Updated README.md accordingly

### CODE REVIEW

File: src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        PersonDTO person = aPersonWithChildrenUnderFour(2);
+        HMRCEligibilityRequest request = anHMRCEligibilityRequestWithPerson(person);
+
+        ResponseEntity<BenefitDTO> benefit = restTemplate.postForEntity(ENDPOINT, request, BenefitDTO.class);
+
+        assertSuccessfulNumberOfChildrenResponse(benefit, 0, 2);
+    }
+
+    @Test
+    void shouldReturnNoMatchForMatchingNino() {
```
Comment (YetAnotherMatt) [inline src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java]: shouldReturnNoMatchForUnmatchedNino

File: src/main/java/uk/gov/dhsc/htbhf/smartstub/model/HMRCEligibilityRequest.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+@Data
+@Builder
+@AllArgsConstructor(onConstructor_ = {@JsonCreator})
+public class HMRCEligibilityRequest {
```
Comment (dwybourn) [inline src/main/java/uk/gov/dhsc/htbhf/smartstub/model/HMRCEligibilityRequest.java]: We could create an abstract EligibilityRequest class as the only difference between the two is the threshold field.

File: src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        ResponseEntity<BenefitDTO> benefit = restTemplate.postForEntity(ENDPOINT, request, BenefitDTO.class);
+
+        assertSuccessfulStatusResponse(benefit, NOMATCH);
+        assertThat(benefit.getBody().getNumberOfChildrenUnderOne()).isNull();
+        assertThat(benefit.getBody().getNumberOfChildrenUnderFour()).isNull();
+    }
+
+    @Test
+    void shouldReturnSameChildrenUnder1AndChildrenUnder4WhenChildrenUnder1TooHigh() {
+        PersonDTO person = aPersonWithChildren(3, 1);
```
Comment (dwybourn) [inline src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java]: Do we not want to throw an exception this case?

File: src/main/java/uk/gov/dhsc/htbhf/smartstub/model/HMRCEligibilityRequest.java (hunk 2 above)
Comment (mikeSmith08) [reply src/main/java/uk/gov/dhsc/htbhf/smartstub/model/HMRCEligibilityRequest.java]: Yep, good idea.

File: src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java (hunk 3 above)
Comment (YetAnotherMatt) [reply src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java]: I think we're better avoiding 'unexplained' errors in the front end - enter what seems to be a perfectly valid NINO and get a 500 error in response

Comment (mikeSmith08) [reply src/test/java/uk/gov/dhsc/htbhf/smartstub/controller/HMRCBenefitControllerIntegrationTest.java]: This was discussed and done in a PR yesterday, the thought was that this check would restrict too many NINOs from being valid, so we'll be matching the under 4 and under 1 children instead and we now have a single NINO that causes an Exception to be thrown. Full details are in the README.md.

File: src/main/java/uk/gov/dhsc/htbhf/smartstub/model/HMRCEligibilityRequest.java (hunk 2 above)
Comment (mikeSmith08) [reply src/main/java/uk/gov/dhsc/htbhf/smartstub/model/HMRCEligibilityRequest.java]: Actually, having tried this, it gets really messy with JSON and Lombok, so I'm going to leave it as it is.

**Pull request merged**